File: depth_to_normal.py
import numpy as np

def depth_to_normalt(depth):
    # z = 0.001
    # dx = x+1 - x
    # dx = np.concatenate((depth[:, 1:], depth[:, -1:]), axis=1) - depth
    # dy = np.concatenate((depth[1:, :], depth[-1:, :]), axis=0) - depth
    print('hello')
    # output = np.zeros((depth.shape[0], depth.shape[1], 3))
    # grad_appended_one = np.concatenate((np.expand_dims(-dx, axis=2), np.expand_dims(-dy, axis=2), np.expand_dims(np.ones(depth.shape) * z, axis=2)), axis=2)
    # return output

File: depth_to_normal.m
function  image = depth_to_normal(depth, normalize)
    image = zeros(size(depth, 1), size(depth, 2), 3);
    z = 0.002;
    dx = [diff(depth, 1, 2), depth(:, 1) - depth(:, end)];
    dy = [diff(depth, 1, 1); depth(1, :) - depth(end, :)];
    % grad = cat(3, dx, dy, z * ones(size(depth)));
    % norm = sqrt(sum(grad.^2, 3));
    image(:, :, 1) = -dx;
    image(:, :, 2) = -dy;
    image(:, :, 3) = z;
    if normalize == 1
        image_per_pixel_norm = sqrt(sum(image.^2, 3));
        image = image ./ image_per_pixel_norm;
    end


    % python code
    % sx = cv2.Sobel(img, cv2.CV_64F, 1, 0, ksize=7)
    % sy = cv2.Sobel(img, cv2.CV_64F, 0, 1, ksize=7)
    % normal = np.dstack([-sx, sy, np.ones_like(sx)])
    % n = np.linalg.norm(normal, axis=2, keepdims=True)
    % normal = normal / n
    % normal = (normal + 1) / 2
    % return normal

    % matlab code

